mlos_bench/mlos_bench/services/remote/azure/azure_auth.py

```diff
@@ -53,18 +56,60 @@ def __init__(self,
         self._access_token = "RENEW *NOW*"
         self._token_expiration_ts = datetime.datetime.now()  # Typically, some future timestamp.
 
+        # Login as ourselves
+        self._cred = azure_id.AzureCliCredential()
+
+    def _init_sp(self):
+        # Perform this initialization outside of __init__ so that environment loading tests
+        # dont need to specifically mock keyvault interactions out
+
+        # Already logged in as SP
+        if isinstance(self._cred, azure_id.CertificateCredential):
+            return
+
+        check_required_params(
+            self.config, {
+                "spClientId",
+                "keyVaultName",
+                "certName",
+                "tenant",
+            }
+        )
+
+        sp_client_id = self.config["spClientId"]
+        keyvault_name = self.config["keyVaultName"]
+        cert_name = self.config["certName"]
+        tenant_id = self.config["tenant"]
+
+        # Get a client for fetching cert info
+        keyvault_secrets_client = SecretClient(
+            vault_url=f"https://{keyvault_name}.vault.azure.net",
+            credential=self._cred,
+        )
+
+        # The certificate private key data is stored as hidden "Secret" (not Key strangely)
+        #  in PKCS12 format, but we need to decode it.
+        secret = keyvault_secrets_client.get_secret(cert_name)
+        assert secret.value is not None
+        cert_bytes = b64decode(secret.value)
+
+        # Reauthenticate as the service principal.
+        self._cred = azure_id.CertificateCredential(tenant_id=tenant_id, client_id=sp_client_id, certificate_data=cert_bytes)
```

Comment:
Is there an expiration on this portion, or is this sufficient to re-login automatically whenever we call the `get_token` on this?

Reply:
Basically, wondering if the check at the top of `if isinstance(self._cred, CertificateCredential)` is sufficient to avoid re-checking the keyvault repeatedly.

Reply:
@bpkroth If I understand what you mean correctly, the user's `AzureCliCredentials` will expire as usual in ~1 day (50 iterations), but on my side I have observed the SP's `CertificateCredential` last for >3 days with a run of 200 iterations without expiring and running into any errors.

Reply:
Sounds good.  We can adjust later if we run into issues I guess.